mllib/src/main/scala/org/apache/spark/ml/feature/Word2Vec.scala

```diff
@@ -194,6 +232,285 @@ final class Word2Vec @Since("1.4.0") (
+            initializeVector(neu1e)
+
+            val scale = 1.0f / contextIds.length
+
+            // feed forward
+            contextIds.foreach { c =>
+              blas.saxpy(vectorSize, scale, syn0, c * vectorSize, 1, contextVector, 0, 1)
+            }
+
+            generateNegativeSamples(random, word, unigramTable, negSamples, wordIndices)
+
+            wordIndices.view.zipWithIndex.foreach { case (wordId, i) =>
+              blas.scopy(vectorSize, syn1, vectorSize * wordId, 1, l2Vectors, vectorSize * i, 1)
+            }
+
+            val rows = negSamples + 1
+            val cols = vectorSize
+            blas
+              .sgemv("T", cols, rows, 1.0f, l2Vectors, 0, cols, contextVector, 0, 1, 0.0f, gb, 0, 1)
+
+            (0 to gb.length-1).foreach {i =>
+              val v = 1.0f / (1 + math.exp(-gb(i)).toFloat)
+              val err = (negLabels(i) - v) * alpha
+              gb.update(i, err)
+            }
+
+            // update for hidden -> output layer
+            blas.sger(cols, rows, 1.0f, contextVector, 1, gb, 1, hiddenLayerUpdate, cols)
+
+            // update hidden -> output layer, syn1
+            wordIndices.view.zipWithIndex.foreach {case (w, i) =>
+              blas.saxpy(vectorSize,
+                1.0f,
+                hiddenLayerUpdate,
+                i * vectorSize,
+                1,
+                syn1,
+                w * vectorSize,
+                1)
+            }
+
+            // update for word vectors
+            blas.sgemv("N", cols, rows, scale, l2Vectors, 0, cols, gb, 0, 1, 1.0f, neu1e, 0, 1)
+
+            // update input -> hidden layer, syn0
+            contextIds.foreach { i =>
+              blas.saxpy(vectorSize, 1.0f, neu1e, 0, 1, syn0, i * vectorSize, 1)
+            }
+            gb.map(math.abs).sum / alpha
+          }
+          logInfo(s"Partition: $i_, Average Batch Error = ${errors.sum / batchSize}")
+        }
+        Iterator((0, syn0), (1, syn1))
+      }
+
+      val aggedMatrices = partialFits.reduceByKey{case (v1, v2) =>
+        blas.saxpy(vocabSize, 1.0f, v2, 1, v1, 1)
+        v1
+      }.collect
+
+      assert(aggedMatrices.length == 2)
+      val norm = 1.0f / $(numPartitions)
+      aggedMatrices.foreach {case (i, syn) =>
+        blas.sscal(syn.length, norm, syn, 0, 1)
+        if (i == 0) {
+          // copy syn0
+          blas.scopy(syn.length, syn, 0, 1, syn0Global, 0, 1)
+        } else {
+          // copy syn1
+          blas.scopy(syn.length, syn, 0, 1, syn1Global, 0, 1)
+        }
+      }
+      syn0bc.destroy(false)
+      syn1bc.destroy(false)
+      val timePerIteration = (System.nanoTime() - iterationStartTime) / 1e6
+      logInfo(s"Total time taken per iteration: ${timePerIteration} ms")
+    }
+    digitSentences.unpersist()
+    vocabMapbc.destroy()
+    unigramTablebc.destroy()
+
+    new feature.Word2VecModel(vocabMap, syn0Global)
+  }
+
+  private def initializeVector(v: Array[Float], value: Float = 0.0f): Unit = {
```

Comment:
Nit: From what I can tell ``value`` is never actually passed in, so it's always equal to ``0.0f``. Maybe call the method ``zeroVector`` and make that more explicit?